scripts/process_gperf.py

```diff
@@ -0,0 +1,140 @@
+of a pointer address, and then doing memcmp() on a string passed in for
+comparison
+
+We are exclusively working with 4-byte pointer values. This script adjusts
+the generated code so that we work with pointers directly and not strings.
+This saves a considerable amount of space.
+"""
+
+def debug(text):
+    if not args.verbose:
+        return
+    sys.stdout.write(os.path.basename(sys.argv[0]) + ": " + text + "\n")
+
+
+def error(text):
+    sys.stderr.write(os.path.basename(sys.argv[0]) + " ERROR: " + text + "\n")
+    sys.exit(1)
+
+
+def warn(text):
+    sys.stdout.write(os.path.basename(sys.argv[0]) + " WARNING: " + text + "\n")
+
+
+def reformat_str(match_obj):
+    addr_str = match_obj.group(0)
+
+    # Nip quotes
+    addr_str = addr_str[1:-1]
+    addr_vals = [0, 0, 0, 0]
+    ctr = 3
+    i = 0
+
+    while (True):
+        if i >= len(addr_str):
+            break
+
+        if addr_str[i] == "\\":
+            if addr_str[i+1].isdigit():
+                # Octal escape sequence
+                val_str = addr_str[i+1:i+4]
+                addr_vals[ctr] = int(val_str, 8)
+                i += 4
+            else:
+                # Char value that had to be escaped by C string rules
+                addr_vals[ctr] = ord(addr_str[i+1])
+                i += 2
+
+        else:
+            addr_vals[ctr] = ord(addr_str[i])
+            i += 1
+
+        ctr -= 1
+
+    return "(char *)0x%02x%02x%02x%02x" % tuple(addr_vals)
+
+def process_line(line, fp):
+    if line.startswith("#"):
+        fp.write(line)
+        return
+
+    # Set the lookup function to static inline so it gets rolled into
+    # _k_object_find(), nothing else will use it
+    if re.search("struct _k_object [*]$", line):
+        fp.write("static inline " + line)
+        return
+
+    # Replace length lookups with constant len of 4 since we're always
+    # looking at pointers
+    line = re.sub(r'lengthtable[[]key[]]', r'4', line)
+
+    # Empty wordlist entries to have NULLs instead of ""
+    line = re.sub(r'[{]["]["][}]', r'{}', line)
+
+    # Suppress a compiler warning since this table is no longer necessary
+    line = re.sub(r'static unsigned char lengthtable',
+                  r'static unsigned char __unused lengthtable', line)
+
+    # drop all use of register keyword, let compiler figure that out,
+    # we have to do this since we change stuff to take the address of some
+    # parameters
+    line = re.sub(r'register', r'', line)
+
+    # Hashing the address of the string
+    line = re.sub(r"hash [(]str, len[)]",
+            r"hash((const char *)&str, len)", line)
+
+    # Just compare pointers directly instead of using memcmp
+    if re.search("if [(][*]str", line):
+        fp.write("            if (str == s)\n")
+        return
+
+    # Take the strings with the binary information for the pointer values,
+    # and just turn them into pointers
+    line = re.sub(r'["].*["]', reformat_str, line)
```

Comment:
More things that could be done here:

* Remove the `len` parameter from `hash()` and `_k_object_lookup()`
* `len` can be added somewhere in `_k_object_lookup()` as a constant value of 4. This should get rid of some branches inside that function

Reply:
looking at the generated assembly code, the compiler appears to be smart enough to disregard len entirely since both hash and k_object_lookup are now static functions, called only from _k_object_find() which uses a constant parameter (4) for len,

I don't currently see a need for these two steps unless I'm missing something.

For example on x86, all of the code currently reduces to:

Dump of assembler code for function _k_object_find:
   0x00002b84 <+0>:	push   %ebp
   0x00002b85 <+1>:	mov    %esp,%ebp
   0x00002b87 <+3>:	push   %eax
   0x00002b88 <+4>:	mov    0x8(%ebp),%ecx
   0x00002b8b <+7>:	movzbl %cl,%eax
   0x00002b8e <+10>:	movzbl 0x38c0(%eax),%edx
   0x00002b95 <+17>:	xor    %eax,%eax
   0x00002b97 <+19>:	cmp    $0x1,%edx
   0x00002b9a <+22>:	ja     0x2bac <_k_object_find+40>
   0x00002b9c <+24>:	cmp    0x40d000(,%edx,8),%ecx
   0x00002ba3 <+31>:	jne    0x2bac <_k_object_find+40>
   0x00002ba5 <+33>:	lea    0x40d000(,%edx,8),%eax
   0x00002bac <+40>:	leave  
   0x00002bad <+41>:	ret    
End of assembler dump.

Reply:
I've made that comment before looking at the assembly output. My bad. Yeah, those changes are unnecessary.